File: Python/Tries/L_extraCharInStr.py
class ExtraChar:
    def __init__(self):
        self.dp = [-1] * 51  # Initialize dp array with -1 values
        
    def minExtraChar(self, s, dictionary):
        return self.minExtraCharHelper(s, dictionary, 0)
    
    def minExtraCharHelper(self, s, dictionary, i):
        if i == len(s):
            return 0

        if self.dp[i] == -1:
            self.dp[i] = 1 + self.minExtraCharHelper(s, dictionary, i + 1)  # Initialize with one extra character.

            for w in dictionary:
                if s[i:i+len(w)] == w:
                    self.dp[i] = min(self.dp[i], self.minExtraCharHelper(s, dictionary, i + len(w)))  # Update if a word in the dictionary is found.

        return self.dp[i]

File: JavaScript/Dynamic Programming/L_extraCharInStr.js
var minExtraChar = function (s, dictionary) {
    let dict = new Set(dictionary); // Store dictionary in a set for O(1) lookup
    let n = s.length;
    let dp = new Array(n + 1).fill(n); // DP array initialized with maximum extra characters
    dp[0] = 0; // No extra characters for an empty string

    // Iterate through the string
    for (let i = 1; i <= n; i++) {
        // Try all substrings ending at i
        for (let j = 0; j < i; j++) {
            let sub = s.substring(j, i); // Get substring s[j:i]
            if (dict.has(sub)) {
                dp[i] = Math.min(dp[i], dp[j]); // If substring found in dictionary
            }
        }
        dp[i] = Math.min(dp[i], dp[i - 1] + 1); // Consider current character as extra
    }

    return dp[n]; // Return the result from dp[n]
};
